Add builtins related to string replacement

kenfdev opened the pull request:
Added the following functions inside the stdlib of go.

* [Regexp.FindAllStringSubmatch](https://golang.org/pkg/regexp/#Regexp.FindAllStringSubmatch) -> `regex.find_all_string_submatch_n`
* [Replacer.Replace](https://golang.org/pkg/strings/#Replacer.Replace) -> `replacer.replace`

Related Issue: #1804 

I'm not sure about the name of the function 🤔 Seems a bit long but couldn't come up with a better one.

Also, this is my first contribution to the code base. Apologies if I'm missing something.

tsandall commented on topdown/strings.go:
```diff
@@ -227,6 +227,28 @@ func builtinReplace(a, b, c ast.Value) (ast.Value, error) {
+		on[i] = string(s.Value.(ast.String))
+	}
+
+	r := strings.NewReplacer(on...)
```
`strings.NewReplacer` will panic if the list contains an odd number of elements. Let's make the API use a map instead of an array so that this becomes a non-issue. I'd also prefer a map syntactically.

kenfdev replied:
Thank you for the review. Do you mean the input becomes a `map` and converts a `map` to an `array` inside the function and than pass it to `NewReplacer`?

One concern is that, IIUC map's don't have guaranteed orders so depending on the input, the output may become flaky.

Or am I misunderstanding the requested change?
Thank you!

tsandall replied:
I was thinking something like:

```
strings.replace_n("the string input", {"th": "!!", "n": "&&"})
```

This would produce:

```
"!!e stri&&g i&&put"
```

> One concern is that, IIUC map's don't have guaranteed orders so depending on the input, the output may become flaky.

I see. I didn't realize that the replacer worked this way. I don't have too strong of an opinion here except that it's unclear what the built-in should do if there are an odd number of arguments. The options are:

1. Return an error -- this is not great because errors are unrecoverable
2. Be undefined
3. Drop the last argument

If we use a map then it's impossible for this to happen.

kenfdev replied:
Maybe I'm being too cautious here, but suppose there was a sting like this

```bash
c/r/u/d
```

and you want to replace it to 

```bash
create/read/update/delete
```

In this case, the replacing order matters. It will succeed in the following order:

1. d -> delete
2. r -> read
3. c -> create
4. u -> update

If you accidentally replace it starting from `c -> create`, the `r` gets replaced to `read` in the latter replacements (`create` becomes `creadeate`). This may lead to problems in the future.

>  it's unclear what the built-in should do if there are an odd number of arguments. The options are:

I agree. I'd go with option 3 `Drop the last argument`. This will need to be well documented but will be `the best it can do` solution. I don't have a strong background for this opinion though.

Option 1 sounds like the best style but not being able to recover seems too much of a disadvantage. Option 2 I think will lead to **hard to debug** undefined rules.

WDYT?

patrick-east replied:
I think we're ok on this. Looking at the source for the replacer it looks like it only iterates over the original string a single time and replaces substrings as it encounters them.

kenfdev replied:
@tsandall @patrick-east Terribly sorry about the confusion. I was somehow stereotyped with the `Replacements are performed in the order they appear in the target string` comment in the go docs. I've understood I was just misunderstanding it.

Should've experimented it myself before saying the wrong thing 😞 

I'll change the arguments to maps and thank you for the reviews.

tsandall commented on ast/builtins.go:
```diff
@@ -766,6 +783,20 @@ var Replace = &Builtin{
+// Replacements are performed in the order they appear in the target string, without overlapping matches.
+// The old string comparisons are done in argument order.
+var ReplacerReplace = &Builtin{
+	Name: "replacer.replace",
```
We don't have to match the Go package names 1:1. I'd prefer something like `strings.replace_n` here.

kenfdev replied:
I agree. I'll go with `strings.replace_n`

kenfdev commented on docs/content/policy-reference.md:
```diff
@@ -77,6 +77,7 @@ complex types.
 | <span class="opa-keep-it-together">``output := indexof(string, search)``</span> | ``output`` is the index inside ``string`` where ``search`` first occurs, or -1 if ``search`` does not exist |
 | <span class="opa-keep-it-together">``output := lower(string)``</span> | ``output`` is ``string`` after converting to lower case |
 | <span class="opa-keep-it-together">``output := replace(string, old, new)``</span> | ``output`` is a ``string`` representing ``string`` with all instances of ``old`` replaced by ``new`` |
+| <span class="opa-keep-it-together">``output := strings.replace_n(patterns, string)``</span> | ``patterns`` is an object with old, new string key value pairs (e.g. ``{"old1": "new1", "old2": "new2", ...}``). ``output`` is a ``string`` with all old strings inside ``patterns`` replaced by the new strings |
```
Wondered if it is proper to say `object` here.

kenfdev commented on topdown/strings.go:
```diff
@@ -227,6 +227,36 @@ func builtinReplace(a, b, c ast.Value) (ast.Value, error) {
+	for k, v := range oldnewObj {
+		strVal, ok := v.(string)
+		if !ok {
+			return nil, fmt.Errorf("non-string value included in object: %s", a.String())
```
`builtins.NewOperandTypeErr` didn't seem to make sense here so I've created a new error here.

kenfdev replied:
Understood. I've made the changes here:

https://github.com/open-policy-agent/opa/pull/1825/files#diff-63b9886aafbee520e97e3ddd5138d87cR250

kenfdev commented:
Thank you @tsandall for all your detailed reviews! I've changed the final error comments and squashed the commit. If the CI passes, this PR should be ready to merge.
